Deleted frames

ActiveChooN opened the pull request:
<!---
Copyright (C) 2020-2022 Intel Corporation

SPDX-License-Identifier: MIT
-->

Resolve #4235
Resolve #3000

<!-- Raised an issue to propose your change (https://github.com/opencv/cvat/issues).
It helps to avoid duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [CONTRIBUTION](https://github.com/opencv/cvat/blob/develop/CONTRIBUTING.md)
guide. -->

<!-- Provide a general summary of your changes in the Title above -->

### Motivation and context
<!-- Why is this change required? What problem does it solve? If it fixes an open
issue, please link to the issue here. Describe your changes in detail, add
screenshots. -->

### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable by a reason then ~~explicitly strikethrough~~ the whole
line. If you don't do that github will show an incorrect process for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I submit my changes into the `develop` branch
- [x] I have added a description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
~~- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly~~
- [x] I have added tests to cover my changes
- [x] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2022 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

ActiveChooN commented:
@bsekachev, @azhavoro, could you look at the PR, please?

azhavoro commented:
> @bsekachev, @azhavoro, could you look at the PR, please?

@ActiveChooN It looks like the `deleted_frames` field is not kept in a task's backup/restore archive. Most likely it will be enough to add this field here https://github.com/openvinotoolkit/cvat/blob/develop/cvat/apps/engine/backup.py#L106

ActiveChooN commented:
@azhavoro, fixed.

bsekachev commented:
- I believe the setting "show deleted frames" is bettter to be in player section
- All frames are grayscale:
![bug_1](https://user-images.githubusercontent.com/40690378/153572460-32cd6f31-c7bf-4bcb-a807-cc0182321847.gif)

- If I give a direct command to go to a deleted frame, it goes to non-deleted neighborhood, I am not sure it is expected behaviour.
![bug_2](https://user-images.githubusercontent.com/40690378/153572511-52bf3664-edbb-46df-8863-fabf2813a65d.gif)

- If I remove the latest frame of a job, it is not removed visually (icon to remove is not changed, grayscale effect is not visible). After updating a page, remove effects are applied.
![bug_3](https://user-images.githubusercontent.com/40690378/153572595-d9f7d57f-b9bf-4f1c-bac9-d91576c7f4d1.gif)


- Looks like removing/restoring frames does not update database field updated_date of a task, so, when I dump annotations, then remove a frame, and dump annotations again, I get cached version of dump where the frame is not removed. Maybe the same bug exists for projects likewise. 

- When I try to export a dataset for a task with one removed image (where some annotations exist, I get following error)
![Screenshot_4](https://user-images.githubusercontent.com/40690378/153572737-5bfe71e5-be8e-49f5-91ac-50305869402c.png)


- According to minutes from our latest discussion:
    - What happens if all frames in job are deleted.
      There will be a message about empty job instead of frame and inactivated annotation view
It is not implemented, as I can see.

ActiveChooN commented:
@bsekachev 

> * I believe the setting "show deleted frames" is bettter to be in player section

Done.

> * All frames are grayscale:

Done.

> If I give a direct command to go to a deleted frame, it goes to a non-deleted neighborhood, I am not sure it is expected behavior.

If you want to navigate through it's better to turn on the "Show deleted frames" option. If not, only navigation through non-deleted frames is available.

> * If I remove the latest frame of a job, it is not removed visually (icon to remove is not changed, grayscale effect is not visible). After updating a page, remove effects are applied.

Fixed.

> * Looks like removing/restoring frames does not update database field updated_date of a task, so, when I dump annotations, then remove a frame, and dump annotations again, I get cached version of dump where the frame is not removed. Maybe the same bug exists for projects likewise.

Fixed.

> * When I try to export a dataset for a task with one removed image (where some annotations exist, I get following error)

Fixed.

> * What happens if all frames in job are deleted.

I thought another time and preferred to show deleted frames if no non-deleted frames are available because the state of the empty job differs from the state when all frames are deleted. And also it makes look of all deleting pipeline more consistent, I guess.

bsekachev commented on cvat-core/tests/mocks/dummy-data.mock.js:
```diff
@@ -2662,8 +2681,9 @@ const frameMetaDummyData = {
         image_quality: 50,
         start_frame: 0,
-        stop_frame: 5001,
+        stop_frame: 499,
```
Why did you change this value?

ActiveChooN replied:
Couse `frameDataCache` stores job-related data meta now, so it's job-related data start/stop frames.

bsekachev commented on cvat-ui/src/actions/annotation-actions.ts:
```diff
@@ -1671,3 +1768,85 @@ export function switchNavigationBlocked(navigationBlocked: boolean): AnyAction {
+                type: AnnotationActionTypes.DELETE_FRAME,
+            });
+
+            await jobInstance.frames.delete(frameData.number);
+            const data = await jobInstance.frames.get(frameData.number);
```
Maybe frames.delete should return a new instance?

ActiveChooN replied:
Getting a frame requires a little bit more arguments. I think this variant also looks normal.

bsekachev commented on cvat-ui/src/actions/annotation-actions.ts:
```diff
@@ -1671,3 +1768,85 @@ export function switchNavigationBlocked(navigationBlocked: boolean): AnyAction {
+                type: AnnotationActionTypes.RESTORE_FRAME,
+            });
+
+            await jobInstance.frames.restore(frameNumber);
+            const frameData = await jobInstance.frames.get(frameNumber);
```
Maybe frames.restore should return a new instance?

bsekachev commented on cvat-ui/src/containers/annotation-page/annotation-page.tsx:
```diff
@@ -62,7 +62,7 @@ function mapDispatchToProps(dispatch: any, own: OwnProps): DispatchToProps {
     const searchParams = new URLSearchParams(window.location.search);
     const initialFilters: object[] = [];
-    let initialFrame = 0;
+    let initialFrame = -1;
```
Why -1?

ActiveChooN replied:
Changed to `null`

ActiveChooN commented on cvat-core/src/frames.js:
```diff
@@ -104,10 +100,26 @@
+                    /**
+                     * Start frame of the frame in the job
+                     * @name startFrame
+                     * @type {integer}
```
There is a lot of type annotation like this and only three of `@type {number}` btw. Should other be fixed too?

ActiveChooN replied:
Fixed everywhere

bsekachev replied:
Yes, thx

bsekachev commented on cvat-ui/src/reducers/notifications-reducer.ts:
```diff
@@ -104,6 +104,8 @@ const defaultState: NotificationsState = {
             search: null,
             searchEmptyFrame: null,
+            searchNonDeletedFrame: null,
+            deleteFrame: null,
```
``restoreFrame`` is not defined here

ActiveChooN replied:
Fixed.

bsekachev commented on tests/cypress/integration/actions_objects2/case_delete_frame.js:
```diff
@@ -0,0 +1,71 @@
+
+        it('Change deleted frame visability.', () => {
+            cy.openSettings();
+            cy.get('.cvat-workspace-settings-show-deleted').within(() => {
+                cy.get('[type="checkbox"]').should('not.be.checked').check();
+            });
```
What do you think about adding corresponding cypress command for delete/restore (and by the way add tests to restore)?

ActiveChooN replied:
There is one. This part is only used for changing settings. Test for restoring frame is located at the end of the test.

bsekachev commented on cvat-ui/src/containers/annotation-page/top-bar/top-bar.tsx:
```diff
@@ -414,20 +457,46 @@ class AnnotationTopBarContainer extends React.PureComponent<Props, State> {
             }
-            this.changeFrame(value);
+            if (!showDeletedFrames) {
+                const frameTo = value >= frameNumber ? jobInstance.startFrame : jobInstance.stopFrame;
+                const frameToBackward = value >= frameNumber ? jobInstance.stopFrame : jobInstance.startFrame;
+                searchNonDeletedFrame(value, frameTo, frameToBackward);
```
According to our HLD we should not search not deleted frame if a user wants to get a specific frame using this field? Isn't it?

I also think that there is very similar case with slider above, also do not need to find not deleted frames. But it may be discussable.

bsekachev commented:
> If you want to navigate through it's better to turn on the "Show deleted frames" option. If not, only navigation through non-deleted frames is available.

It does not satisfy discussed HLD. And I actually do not think it is expected behaviour for users because they even does not know about this checkbox exists.

ActiveChooN commented on cvat-canvas/src/typescript/canvasView.ts:
```diff
@@ -515,13 +518,18 @@ export class CanvasViewImpl implements CanvasView, Listener {
 
+        // Transform deleted frame overlay pattern
+        this.deletedImageOverlayPath.setAttribute('stroke-width', `${consts.BASE_DELETED_PATTERN_WIDTH / this.geometry.scale}px`);
+
```
Fixed.

ActiveChooN commented on cvat-canvas/src/typescript/canvasView.ts:
```diff
@@ -1260,6 +1299,21 @@ export class CanvasViewImpl implements CanvasView, Listener {
+
+                if (model.imageDeleted) {
+                    this.deletedImageOverlay.style.display = null;
+                    if (!this.background.hasAttribute('prevFilter')) {
```
Switched to using `configure` method attribute for that purpose.

ActiveChooN commented on cvat-core/src/frames.js:
```diff
@@ -704,48 +726,93 @@
+    async function deleteFrame(jobID, frame) {
+        const { meta } = frameDataCache[jobID];
+        if (meta.deleted_frames.includes(frame)) {
+            throw new ArgumentError('Frame should not be deleted');
```
Fixed

ActiveChooN commented on cvat-core/src/session.js:
```diff
@@ -1895,6 +2004,28 @@
+        return result;
+    };
+
+    Job.prototype.frames.searchNonDeleted.implementation = async function (frameFrom, frameTo) {
+        if (!Number.isInteger(frameFrom) || !Number.isInteger(frameTo)) {
+            throw new ArgumentError('The start and end frames both must be an integer');
+        }
+
+        if (frameFrom < this.startFrame || frameFrom > this.stopFrame) {
+            throw new ArgumentError('The start frame is out of the job');
+        }
+
+        if (frameTo < this.startFrame || frameTo > this.stopFrame) {
+            throw new ArgumentError('The stop frame is out of the job');
+        }
+
+        const result = await searchNonDeletedFrame(this.id, frameFrom, frameTo);
+        return result;
+    };
```
In the first approach, I've tried to extend the current search method. But `search` method operates with annotations only, but some data from `frameDataCache`. We have a separate method for searching empty frames so I don't see a big problem here.

bsekachev replied:
Let me rephrase, it is okay to have a dedicated method to search frames, but I suggest to rename it from: 

``Job.prototype.frames.searchNonDeleted`` to ``Job.prototype.frames.search`` and add additional argument (what to find?): deleted/not deleted

bsekachev replied:
Or at least for now just rename it (without adding an argument) to escape backward compatibility lose in the future

ActiveChooN replied:
> I suggest to rename it from:
> 
> `Job.prototype.frames.searchNonDeleted` to `Job.prototype.frames.search`

Ok, I've got your point. Sounds relevant.

ActiveChooN replied:
Fixed.

ActiveChooN commented on cvat-core/src/session.js:
```diff
@@ -1628,6 +1694,23 @@
+                            }
+                            data.deletedFrames = frames;
+                        },
+                    },
```
Deleted

ActiveChooN commented on cvat-core/src/session.js:
```diff
@@ -2232,6 +2368,64 @@
+        const jobs = this.jobs.filter((_job) => (
+            (frameFrom >= _job.startFrame && frameFrom <= _job.stopFrame) &&
+            (frameTo >= _job.startFrame && frameTo <= _job.stopFrame) &&
+            (frameFrom < _job.startFrame && frameTo > _job.stopFrame)
```
Of course, it should be concatenated with `||`, my fault.

ActiveChooN commented on cvat-ui/src/actions/annotation-actions.ts:
```diff
@@ -1017,7 +1027,15 @@ export function getJobAsync(tid: number, jid: number, initialFrame: number, init
 
-            const frameNumber = Math.max(Math.min(job.stopFrame, initialFrame), job.startFrame);
+            let frameNumber;
+            if (initialFrame < 0 && !showDeletedFrames) {
```
It's just a marker, that no frame was selected with a URL. Switched to using a more understandable `null`.

ActiveChooN commented on cvat-ui/src/actions/annotation-actions.ts:
```diff
@@ -1387,10 +1405,35 @@ export function changeGroupColorAsync(group: number, color: string): ThunkAction
+                if (frame !== searchedFrame) {
+                    return;
+                }
+            }
```
Yeap, it looks much simpler, thanks.

ActiveChooN commented:
@bsekachev, @nmanovic, PR is mostly done, the only issue with interpolating on the client. If the job has keyframes on deleted frames there can be an annotation shift on nearby frames, if those keyframes are ignored in the interpolation process. That issue is not related to interpolation on the server, because in that situation keyframes are deleted completely.

bsekachev commented:
> Update overlay filter for deleted frames for more clarity;

As far as I remember for deleted frames we discussed something like trash icon.. What do you think?

I found one issue with step forward/backward navigation: If I have deleted frames 0,1,2 and go to the 3rd frame, press backward, backward, backward I visite frames 2,1,0. If I press back, back, back navigation don't change frame (as expected). Need to check the same case when we have deleted frames latest, latest - 1, latest - 2 and go forward. 

> Add keyframes at the edges of the deleted interval to reduce annotations shift;

I have a polygon track with three keyframes (0, 10, 11 [outside keyframe]), I dump data as CVAT for video and don not see any edge keyframes on frame 1. Did I do something wrong?

bsekachev commented:
> If the job has keyframes on deleted frames there can be an annotation shift on nearby frames, if those keyframes are ignored in the interpolation process.

As far as I remember we discussed that **we can use** information about keyframes on removed frames during interpolation on client, but maybe it is not exactly true as I think now.

Maybe better idea is to add keyframes automatically around frame that has been just removed? For example if we remove frame N, we add for all tracks (on frame N) keyframes on frames N-1 (if exists, and the track is presented there,  and not keyframe so far), and N+1 (if exists and the track is presented there,  and not keyframe so far), keeping position as is. It corresponds to the server solution, and not break any data somehow (maybe it a bit less convenient when we want to restore removed frame, but I am not sure it is really often scenario).

Any ideas?

bsekachev commented:
BTW, are we planning to do something with filters in this PR? (search deleted frames, navigation redesign, etc.?) considering Andrey's feedback https://github.com/openvinotoolkit/cvat/issues/3075#issuecomment-1047589431?

bsekachev commented:
If I have a track with all keyframes on deleted frames (including outside keyframes), I get dataset export error: List index out of range.

bsekachev commented:
One more related issue with interpolation (inconsistency). If I remove a frame with outside track keyframe, on UI I don't see the track anymore (as expected) on the following frames, but in dumped file the track lasts till the end of a video.

nmanovic commented:
@ActiveChooN , need to resolve conflicts

klakhov commented:
Tried this patch: deleting frames, import/export with deleted frames works fine for me. Didn't meet any issues

azhavoro commented on docker-compose.yml:
```diff
@@ -13,6 +13,8 @@ services:
       POSTGRES_DB: cvat
       POSTGRES_HOST_AUTH_METHOD: trust
+    ports:
+      - 5432:5432
```
Why this change is needed?

ActiveChooN replied:
Fixed

nmanovic commented on site/content/en/docs/manual/advanced/delete_frame.md:
```diff
@@ -0,0 +1,50 @@
+   ![](/images/image245.jpg)
+
+1. After that you will be asked to confirm frame deleting.
+   **Important note:**  all annotaiton from that frame will be deleted, unsaved annotations
```
@ActiveChooN , there is a typo in the sentence but I'm not sure I understand the statement about "unsaved annotations". Could you please explain?

ActiveChooN replied:
We need to flush unsaved annotations before clearing frame annotations and deleting the frame.

nmanovic commented on site/content/en/docs/manual/advanced/delete_frame.md:
```diff
@@ -0,0 +1,50 @@
+   will be saved and frame will be unvisable in the annotation view (Until you make it visible in the settings).
+   If there is some overlap in the task and the deleted frame falls within this interval,
+   then this will cause this frame to become unavailable in another job as well.
+1. If you annotate interpolation task, you may need to make some correction in the annotation. Common ones are:
```
When you delete a frame in a job with tracks, you may need adjust some tracks if the deleted frame has key frames for these tracks. You will need to adjust these tracks manually.

ActiveChooN replied:
Changed. But it also can relate with any deleted frames not only ones with keyframes.

nmanovic commented on site/content/en/docs/manual/advanced/delete_frame.md:
```diff
@@ -0,0 +1,50 @@
+1. Then you will be able to navigate through deleted frames.
+   But annotation tools will be unavailable. Deleted frames differ in the corresponding overlay.
+
+# Restore deleted frame
```
I cannot find information about navigation. As we discussed there are several ways to go to a deleted frames directly (without adjusting settings). Could you please add a navigation section?

ActiveChooN replied:
Added

nmanovic commented on cvat/apps/engine/views.py:
```diff
@@ -1148,6 +1158,77 @@ def data(self, request, pk):
+    @extend_schema(summary='Method provides a meta information about media files which are related with the job',
+        responses={
+            '200': DataMetaSerializer,
+        }, tags=['tasks'], versions=['2.0'])
```
@Marishka17 , could you please look at the definition? I remember that in your patch all tags and versions were removed.

Marishka17 replied:
Yes, it should be removed

ActiveChooN replied:
Fixed

nmanovic commented on cvat/apps/engine/tests/test_rest_api.py:
```diff
@@ -399,6 +399,49 @@ def test_api_v2_jobs_id_admin_partial(self):
         response = self._run_api_v2_jobs_id(self.job.id, self.owner, data)
         self._check_request(response, data)
 
+class JobDataMetaPartialUpdateAPITestCase(APITestCase):
```
@ActiveChooN , can we implement these tests in tests/rest_api instead? These files should not be extended anymore.

nmanovic commented on cvat/apps/iam/permissions.py:
```diff
@@ -789,6 +790,8 @@ def get_scopes(request, view, obj):
             ('annotations', 'DELETE'): 'delete:annotations',
             ('annotations', 'PUT'): 'update:annotations',
             ('data', 'GET'): 'view:data',
+            ('data_info','GET'): 'view:data',
```
`view:data`, `update:data` to `view:metadata` and `update:metadata`
I will recommend to rename data_info to metadata

ActiveChooN replied:
Changed

ActiveChooN replied:
Cannot rename `data_info` to `metadata` due to this check
<https://github.com/openvinotoolkit/cvat/blob/develop/cvat/apps/iam/permissions.py#L1041:L1042>

nmanovic replied:
@ActiveChooN , can we make the check more strict? For example, check view.basename?

ActiveChooN replied:
Basename would be task or jobs in that case. But I don't really know why is this check even needed, can't find any mention of `metadata` in any view. I suggest simply delete second part of the check.

nmanovic replied:
https://www.django-rest-framework.org/api-guide/metadata/

ActiveChooN commented on cvat/apps/iam/rules/tasks.csv:
```diff
@@ -60,6 +60,10 @@ view:data,Task,Sandbox,None,,GET,"/tasks/{id}/data, /tasks/{id}/data/meta",Admin
 view:data,Task,Sandbox,"Owner, Project:owner, Assignee, Project:assignee",,GET,"/tasks/{id}/data, /tasks/{id}/data/meta",None,N/A
 view:data,Task,Organization,None,,GET,"/tasks/{id}/data, /tasks/{id}/data/meta",User,Maintainer
 view:data,Task,Organization,"Owner, Project:owner, Assignee, Project:assignee",,GET,"/tasks/{id}/data, /tasks/{id}/data/meta",None,Worker
+update:data,Task,Sandbox,None,,PATCH,/tasks/{id}/data/meta,Admin,N/A
```
Fixed

ActiveChooN commented on cvat/apps/iam/rules/utils.rego:
```diff
@@ -42,6 +42,7 @@ DELETE_ANNOTATIONS := "delete:annotations"
 VIEW_DATA := "view:data"
 VIEW_COMMITS := "view:commits"
 UPLOAD_DATA := "upload:data"
+UPDATE_DATA := "update:data"
```
Fixed

nmanovic commented on cvat/apps/iam/permissions.py:
```diff
@@ -1039,7 +1042,7 @@ def has_object_permission(self, request, view, obj):
     @staticmethod
     def is_metadata_request(request, view):
-        return request.method == 'OPTIONS' or view.action == 'metadata'
+        return request.method == 'OPTIONS'
```
@ActiveChooN , could you please fix?

nmanovic commented on site/content/en/docs/manual/advanced/delete_frame.md:
```diff
@@ -0,0 +1,56 @@
+
+1. After that you will be asked to confirm frame deleting.
+   **Important note:**  all annotations from that frame will be deleted, unsaved annotations
+   will be saved and the frame will be invisible in the annotation view (Until you make it visible in the settings).
```
@ActiveChooN , "unsaved annotations will be saved". Could you please explain? Why do we need that?

ActiveChooN replied:
Unsaved annotations related to the deleting frame can be there. So, as we clear all annotations from the deleted frame, it's better to save annotations before.

nmanovic replied:
@ActiveChooN , I don't understand why we need to save annotations in the case to be honest. For me it isn't consistent with current behaviors.

ActiveChooN replied:
Because deleting a frame also deletes all annotations on that frame. So, there can be an unsaved annotation on the frame and we need to save it first. I will try to investigate to reduce saving if there are no annotations on the frame.

nmanovic commented:
Go to http://localhost:8080/tasks?page=1:

```console
022-03-21 08:08:46,998 DEBG 'runserver' stderr output:
[Mon Mar 21 08:08:46.997879 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340] ERROR - 2022-03-21 08:08:46,997 - log - Internal Server Error: /api/tasks/12/data
[Mon Mar 21 08:08:46.997881 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340] Traceback (most recent call last):
[Mon Mar 21 08:08:46.997882 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
[Mon Mar 21 08:08:46.997883 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     response = get_response(request)
[Mon Mar 21 08:08:46.997883 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
[Mon Mar 21 08:08:46.997884 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     response = wrapped_callback(request, *callback_args, **callback_kwargs)
[Mon Mar 21 08:08:46.997885 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
[Mon Mar 21 08:08:46.997886 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     return view_func(*args, **kwargs)
[Mon Mar 21 08:08:46.997887 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/viewsets.py", line 125, in view
[Mon Mar 21 08:08:46.997887 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     return self.dispatch(request, *args, **kwargs)
[Mon Mar 21 08:08:46.997888 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/views.py", line 509, in dispatch
[Mon Mar 21 08:08:46.997889 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     response = self.handle_exception(exc)
[Mon Mar 21 08:08:46.997890 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/views.py", line 469, in handle_exception
[Mon Mar 21 08:08:46.997891 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     self.raise_uncaught_exception(exc)
[Mon Mar 21 08:08:46.997891 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
[Mon Mar 21 08:08:46.997892 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     raise exc
[Mon Mar 21 08:08:46.997893 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/views.py", line 506, in dispatch
[Mon Mar 21 08:08:46.997893 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     response = handler(request, *args, **kwargs)
[Mon Mar 21 08:08:46.997894 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/home/django/cvat/apps/engine/views.py", line 757, in data
[Mon Mar 21 08:08:46.997895 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     return data_getter(request, self._object.data.start_frame,
[Mon Mar 21 08:08:46.997896 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340] AttributeError: 'NoneType' object has no attribute 'start_frame'

```

nmanovic commented:
Go to http://localhost:8080/tasks?page=1:

```console
022-03-21 08:08:46,998 DEBG 'runserver' stderr output:
[Mon Mar 21 08:08:46.997879 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340] ERROR - 2022-03-21 08:08:46,997 - log - Internal Server Error: /api/tasks/12/data
[Mon Mar 21 08:08:46.997881 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340] Traceback (most recent call last):
[Mon Mar 21 08:08:46.997882 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
[Mon Mar 21 08:08:46.997883 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     response = get_response(request)
[Mon Mar 21 08:08:46.997883 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
[Mon Mar 21 08:08:46.997884 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     response = wrapped_callback(request, *callback_args, **callback_kwargs)
[Mon Mar 21 08:08:46.997885 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
[Mon Mar 21 08:08:46.997886 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     return view_func(*args, **kwargs)
[Mon Mar 21 08:08:46.997887 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/viewsets.py", line 125, in view
[Mon Mar 21 08:08:46.997887 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     return self.dispatch(request, *args, **kwargs)
[Mon Mar 21 08:08:46.997888 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/views.py", line 509, in dispatch
[Mon Mar 21 08:08:46.997889 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     response = self.handle_exception(exc)
[Mon Mar 21 08:08:46.997890 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/views.py", line 469, in handle_exception
[Mon Mar 21 08:08:46.997891 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     self.raise_uncaught_exception(exc)
[Mon Mar 21 08:08:46.997891 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
[Mon Mar 21 08:08:46.997892 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     raise exc
[Mon Mar 21 08:08:46.997893 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/views.py", line 506, in dispatch
[Mon Mar 21 08:08:46.997893 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     response = handler(request, *args, **kwargs)
[Mon Mar 21 08:08:46.997894 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]   File "/home/django/cvat/apps/engine/views.py", line 757, in data
[Mon Mar 21 08:08:46.997895 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340]     return data_getter(request, self._object.data.start_frame,
[Mon Mar 21 08:08:46.997896 2022] [wsgi:error] [pid 599:tid 139669365364480] [remote 172.19.0.4:55340] AttributeError: 'NoneType' object has no attribute 'start_frame'

```

nmanovic commented:
http://localhost:8080/api/jobs/1/data/meta

Server Error (500)

```console
022-03-21 08:16:58,214 DEBG 'runserver' stderr output:
[Mon Mar 21 08:16:58.214491 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312] [2022-03-21 08:16:58,214] ERROR django.request: Internal Server Error: /api/jobs/1/data/meta
[Mon Mar 21 08:16:58.214501 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312] Traceback (most recent call last):
[Mon Mar 21 08:16:58.214502 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
[Mon Mar 21 08:16:58.214503 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     response = get_response(request)
[Mon Mar 21 08:16:58.214504 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 204, in _get_response
[Mon Mar 21 08:16:58.214504 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     response = response.render()
[Mon Mar 21 08:16:58.214505 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/django/template/response.py", line 105, in render
[Mon Mar 21 08:16:58.214506 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     self.content = self.rendered_content
[Mon Mar 21 08:16:58.214507 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/response.py", line 70, in rendered_content
[Mon Mar 21 08:16:58.214508 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     ret = renderer.render(self.data, accepted_media_type, context)
[Mon Mar 21 08:16:58.214509 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/renderers.py", line 724, in render
[Mon Mar 21 08:16:58.214509 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     context = self.get_context(data, accepted_media_type, renderer_context)
[Mon Mar 21 08:16:58.214510 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/renderers.py", line 657, in get_context
[Mon Mar 21 08:16:58.214511 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     raw_data_patch_form = self.get_raw_data_form(data, view, 'PATCH', request)
[Mon Mar 21 08:16:58.214512 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/renderers.py", line 541, in get_raw_data_form
[Mon Mar 21 08:16:58.214513 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     if not self.show_form_for_method(view, method, request, instance):
[Mon Mar 21 08:16:58.214514 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/renderers.py", line 433, in show_form_for_method
[Mon Mar 21 08:16:58.214514 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     view.check_object_permissions(request, obj)
[Mon Mar 21 08:16:58.214515 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/views.py", line 345, in check_object_permissions
[Mon Mar 21 08:16:58.214516 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     if not permission.has_object_permission(request, self, obj):
[Mon Mar 21 08:16:58.214517 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/home/django/cvat/apps/iam/permissions.py", line 1041, in has_object_permission
[Mon Mar 21 08:16:58.214518 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     return self.check_permission(request, view, obj)
[Mon Mar 21 08:16:58.214519 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/home/django/cvat/apps/iam/permissions.py", line 1030, in check_permission
[Mon Mar 21 08:16:58.214520 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     permissions.extend(perm.create(request, view, obj))
[Mon Mar 21 08:16:58.214521 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/home/django/cvat/apps/iam/permissions.py", line 762, in create
[Mon Mar 21 08:16:58.214522 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     self = cls.create_base_perm(request, view, scope, obj)
[Mon Mar 21 08:16:58.214523 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/home/django/cvat/apps/iam/permissions.py", line 21, in create_base_perm
[Mon Mar 21 08:16:58.214523 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     return cls(
[Mon Mar 21 08:16:58.214524 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/home/django/cvat/apps/iam/permissions.py", line 777, in __init__
[Mon Mar 21 08:16:58.214525 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     super().__init__(**kwargs)
[Mon Mar 21 08:16:58.214526 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/home/django/cvat/apps/iam/permissions.py", line 71, in __init__
[Mon Mar 21 08:16:58.214527 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     self.payload['input']['resource'] = self.get_resource()
[Mon Mar 21 08:16:58.214527 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/home/django/cvat/apps/iam/permissions.py", line 838, in get_resource
[Mon Mar 21 08:16:58.214528 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     if self.obj.segment.task.project:
[Mon Mar 21 08:16:58.214529 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312] AttributeError: 'Data' object has no attribute 'segment'

2022-03-21 08:16:58,214 DEBG 'runserver' stderr output:
[Mon Mar 21 08:16:58.214597 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312] ERROR - 2022-03-21 08:16:58,214 - log - Internal Server Error: /api/jobs/1/data/meta
[Mon Mar 21 08:16:58.214599 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312] Traceback (most recent call last):
[Mon Mar 21 08:16:58.214599 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
[Mon Mar 21 08:16:58.214600 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     response = get_response(request)
[Mon Mar 21 08:16:58.214601 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 204, in _get_response
[Mon Mar 21 08:16:58.214602 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     response = response.render()
[Mon Mar 21 08:16:58.214602 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/django/template/response.py", line 105, in render
[Mon Mar 21 08:16:58.214603 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     self.content = self.rendered_content
[Mon Mar 21 08:16:58.214604 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/response.py", line 70, in rendered_content
[Mon Mar 21 08:16:58.214605 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     ret = renderer.render(self.data, accepted_media_type, context)
[Mon Mar 21 08:16:58.214606 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/renderers.py", line 724, in render
[Mon Mar 21 08:16:58.214606 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     context = self.get_context(data, accepted_media_type, renderer_context)
[Mon Mar 21 08:16:58.214607 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/renderers.py", line 657, in get_context
[Mon Mar 21 08:16:58.214608 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     raw_data_patch_form = self.get_raw_data_form(data, view, 'PATCH', request)
[Mon Mar 21 08:16:58.214609 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/renderers.py", line 541, in get_raw_data_form
[Mon Mar 21 08:16:58.214609 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     if not self.show_form_for_method(view, method, request, instance):
[Mon Mar 21 08:16:58.214610 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/renderers.py", line 433, in show_form_for_method
[Mon Mar 21 08:16:58.214611 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     view.check_object_permissions(request, obj)
[Mon Mar 21 08:16:58.214612 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/opt/venv/lib/python3.8/site-packages/rest_framework/views.py", line 345, in check_object_permissions
[Mon Mar 21 08:16:58.214613 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     if not permission.has_object_permission(request, self, obj):
[Mon Mar 21 08:16:58.214613 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/home/django/cvat/apps/iam/permissions.py", line 1041, in has_object_permission
[Mon Mar 21 08:16:58.214614 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     return self.check_permission(request, view, obj)
[Mon Mar 21 08:16:58.214615 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/home/django/cvat/apps/iam/permissions.py", line 1030, in check_permission
[Mon Mar 21 08:16:58.214619 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     permissions.extend(perm.create(request, view, obj))
[Mon Mar 21 08:16:58.214620 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/home/django/cvat/apps/iam/permissions.py", line 762, in create
[Mon Mar 21 08:16:58.214621 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     self = cls.create_base_perm(request, view, scope, obj)
[Mon Mar 21 08:16:58.214622 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/home/django/cvat/apps/iam/permissions.py", line 21, in create_base_perm
[Mon Mar 21 08:16:58.214622 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     return cls(
[Mon Mar 21 08:16:58.214623 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/home/django/cvat/apps/iam/permissions.py", line 777, in __init__
[Mon Mar 21 08:16:58.214624 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     super().__init__(**kwargs)
[Mon Mar 21 08:16:58.214625 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/home/django/cvat/apps/iam/permissions.py", line 71, in __init__
[Mon Mar 21 08:16:58.214625 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     self.payload['input']['resource'] = self.get_resource()
[Mon Mar 21 08:16:58.214626 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]   File "/home/django/cvat/apps/iam/permissions.py", line 838, in get_resource
[Mon Mar 21 08:16:58.214627 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312]     if self.obj.segment.task.project:
[Mon Mar 21 08:16:58.214628 2022] [wsgi:error] [pid 608:tid 140081950660352] [remote 172.18.0.5:58312] AttributeError: 'Data' object has no attribute 'segment'
```

nmanovic commented:
@ActiveChooN , after I delete an image and restore it, some frames have unexpected white boundaries. It looks like wrong meta information is used to calculate the size of images on the canvas.

![image](https://user-images.githubusercontent.com/40690625/159551727-98bb997c-02f7-40bf-8b04-6207158324ec.png)

nmanovic commented:
@ActiveChooN , after I delete an image and restore it, some frames have unexpected white boundaries. It looks like wrong meta information is used to calculate the size of images on the canvas.

![image](https://user-images.githubusercontent.com/40690625/159551727-98bb997c-02f7-40bf-8b04-6207158324ec.png)

nmanovic commented:
@ActiveChooN , I believe we decided to draw "trash bin" for deleted images instead of strips. Could you please clarify?

![image](https://user-images.githubusercontent.com/40690625/159552169-ee339405-7a04-49d9-b165-e3151bf747be.png)

nmanovic commented:
@ActiveChooN , I don't think we need to show the dialog, if we don't have any annotations on the frame

![image](https://user-images.githubusercontent.com/40690625/159552835-955d206e-a3ba-4785-a281-72344c51d0fb.png)

nmanovic commented:
@ActiveChooN , I don't think we need to show the dialog, if we don't have any annotations on the frame

![image](https://user-images.githubusercontent.com/40690625/159552835-955d206e-a3ba-4785-a281-72344c51d0fb.png)

nmanovic commented:
@ActiveChooN , after I click once on "delete" button, I got the following requests below. It looks like we do requests twice. Why? Also I will prefer don't send any requests to the server and do that explicitly only after "save" button is pressed. It will correspond to the current behaviour.

![image](https://user-images.githubusercontent.com/40690625/159553783-bb6d6c4a-479f-4e79-a7f1-d30cdb254509.png)

ActiveChooN commented:
> @ActiveChooN , after I delete an image and restore it, some frames have unexpected white boundaries. It looks like wrong meta information is used to calculate the size of images on the canvas.

@nmanovic, can you please share task backup for issue reproduction?

ActiveChooN commented:
> @ActiveChooN , I don't think we need to show the dialog, if we don't have any annotations on the frame.

In my opinion, every major delete action should be wrapped into dialog to prevent missclicking

nmanovic commented:
> > @ActiveChooN , after I delete an image and restore it, some frames have unexpected white boundaries. It looks like wrong meta information is used to calculate the size of images on the canvas.
> 
> @nmanovic, can you please share task backup for issue reproduction?

Please use https://public.roboflow.com/object-detection/oxford-pets to reproduce. It is important just have images with different sizes. Let me know if you cannot reproduce.

nmanovic commented:
> > @ActiveChooN , I don't think we need to show the dialog, if we don't have any annotations on the frame.
> 
> In my opinion, every major delete action should be wrapped into dialog to prevent missclicking

Only in case it has strong consequences. Also I don't like that we move to the next image automatically. Need to think twice about the behaviors.

ActiveChooN commented:
> Let me know if you cannot reproduce.

Should be there special steps to reproduce the issue?

![delete-resizes-image](https://user-images.githubusercontent.com/8168727/159683657-9a1bdc2e-072b-45c9-b631-131e290705b3.gif)

ActiveChooN commented:
@nmanovic 

> Also I don't like that we move to the next image automatically. Need to think twice about the behaviors.

Moving to the next frame is only working if the "Show deleted frames" option is disabled.

nmanovic commented:
@ActiveChooN , automatic annotation is running on all frames, even deleted.

nmanovic commented:
@ActiveChooN , could you please check tests? Are they hanging?

nmanovic commented:
@ActiveChooN , if I delete a frame with annotations and after that restore it, I still can see annotations. But if I go to the next frame and return back, it is disappeared. Could you please check?

nmanovic commented:
@ActiveChooN , do we have any shortcuts to delete a frame quickly? For now it is necessary for every deleted frame click twice using mouse.  I will have two shortcuts with and without confirmation and show them in hints.

nmanovic commented:
@ActiveChooN , don't like that in meta information we have a separate array for deleted frames. Do you see any ways to have "is_deleted" attribute as a part of information about a frame? 

![image](https://user-images.githubusercontent.com/40690625/161277174-134f076c-879c-403e-8f5f-f8f6a561d7f5.png)

nmanovic commented:
@ActiveChooN , why do we have a warning when we delete a frame without annotation but don't have a warning when we restore the frame?

nmanovic commented:
@ActiveChooN , found a problem with tracks. Let's draw a track with key frames on 0, 10, 20 frames. If you delete the frame 0, the track will be removed completely. It tells me that the idea to delete annotations could be wrong and dangerous. Need to think one more time about that. Only in CVAT format we dump key frames only. It looks like it is the main problem.

ActiveChooN commented:
@nmanovic,

> @ActiveChooN , don't like that in meta information we have a separate array for deleted frames. Do you see any ways to have "is_deleted" attribute as a part of information about a frame?

We have video meta without object per frame information, so `is_deleted` key is not applicable in that case.



> @ActiveChooN , I see some strange artifacts for trash bin. Is it a standard picture?

It's the default delete icon in antd, fixed it manually.

violino22 commented:
"It will be possible to mark a frame as deleted after the PR is merged"
What is "PR is merged"?

nmanovic commented:
The pull request (PR) should be merged into develop branch and after that we should publish a release.

violino22 commented:
> Запрос на включение (PR) должен быть объединен с веткой разработки, после чего мы должны опубликовать релиз.

Я просто наёмный работник)))
Я не хочу заморачиваться. Я просто хочу клацать на кнопочки, чтоб удалялись ненужные файлы из задачи, это возможно?)))

nmanovic commented:
Yes, I understand that. In this case just wait. I hope in next release the feature should be available.

FrankUSC commented:
@nmanovic 
This is a very useful feature, do we have an ETA when will it be released?
Thank you!

nmanovic commented:
@FrankUSC , it will be release this month.

bsekachev commented:
The pull request has been finished and merged in our current fork, here: https://github.com/cvat-ai/cvat
